---
title: "Elastic Net and K-Nearest Neighbor Breast Cancer Classification using TCGA DNA Methylation data"
author: "Samuel Nkrumah"
date: "9th September 2021"
output: html_document
---
<style type="text/css">
  body{
 font-size: 12pt;
}
</style>

# 1 Load appropriate packages
```{r eval= FALSE}
library('TCGAbiolinks')
library('SummarizedExperiment')
library("factoextra")
library('limma')
library('edgeR')
library('caret')
library('glmnet')
library('survival')
library('gplots')
library('survminer')
library('stringr')
library("RColorBrewer")
library("genefilter")
library('class')
library('e1071')
```
<br></br>

# 2.1 Query GDC for BRCA DNA Methylation and Gene Expression Data
```{r eval=FALSE}
brca_DNA_Meth_query <- GDCquery(
  project = "TCGA-BRCA",
  data.category = "DNA Methylation",
  data.type = 'Methylation Beta Value',
  platform = 'Illumina Human Methylation 27',
  sample.type = c('Primary Tumor', 'Solid Tissue Normal'))


brca_RNA_query <- GDCquery(
  project = 'TCGA-BRCA',
  data.category = 'Transcriptome Profiling',
  experimental.strategy = 'RNA-Seq',
  workflow.type = 'HTSeq - Counts',
  sample.type = c('Primary Tumor', 'Solid Tissue Normal'))
```

# 2.2 Get and Download DNA methylation data
- Get common patients in both Methylation and Gene Expression Data

```{r eval=FALSE}
DNA_Meth_cases <- substr(getResults(brca_DNA_Meth_query, cols = 'cases'), 1, 12)
RNA_cases <- substr(getResults(brca_RNA_query, cols = 'cases'), 1, 12)
common_patients <- Reduce(intersect, list(DNA_Meth_cases, RNA_cases))


new_brca_DNA_Meth_query <- GDCquery(
  project = "TCGA-BRCA",
  data.category = "DNA Methylation",
  data.type = 'Methylation Beta Value',
  platform = 'Illumina Human Methylation 27',
  sample.type = c('Primary Tumor', 'Solid Tissue Normal'),
  barcode = common_patients[1:100])

GDCdownload(query = new_brca_DNA_Meth_query)
DNA_meth_data <- GDCprepare(new_brca_DNA_Meth_query)
```
<br></br>

# 3.1 Get training data
- Get and transpose DNA_methylation beta values. Omit NA values
```{r eval=FALSE}
beta_values <- assay(DNA_meth_data)
beta_values <- na.omit(beta_values)
beta_values <- t(beta_values)

sample_types <- as.factor(DNA_meth_data$definition)

set.seed(42)
train_ids <- createDataPartition(sample_types, p=0.75, list=FALSE)

x_train <- beta_values[train_ids, ]
x_test <- beta_values[-train_ids, ]

y_train <- sample_types[train_ids]
y_test <- sample_types[-train_ids]
```

# 3.2 Build Elastic Net model with bata values. Make predictions about sample type
```{r eval=FALSE}
EN_model <- cv.glmnet(
  x = x_train,
  y = y_train,
  alpha = 0.5,
  family = 'binomial'
)

EN_y_pred <- predict(EN_model, newx = x_test, type = 'class', s = 'lambda.min')
```

# 3.3 Confusion matrix to check for model accuracy
```{r eval=FALSE}
confusion_matrix = table(EN_y_pred, y_test)
print(confusion_matrix)
```

- Precision of EN model: 1
- Specificity of EN model: 1
- Sensitivity of EN model: 0.96
```{r eval=FALSE}
print(paste0('Precision :', precision(confusion_matrix)))

print(paste0('Specificity :', specificity(confusion_matrix)))

print(paste0('Sensitivity :', sensitivity(confusion_matrix)))
```

# 3.4 Filter beta values and retain Elastic Net Model. Make new predictions
```{r eval=FALSE}

fil_beta_values <- varFilter(t(beta_values), var.func=IQR, var.cutoff=0.95)
fil_beta_values <- t(fil_beta_values)

new_x_train <- fil_beta_values[train_ids,]
new_x_test <- fil_beta_values[-train_ids,]

new_EN_model <- cv.glmnet(
  x = new_x_train,
  y = y_train,
  alpha = 0.5,
  family = 'binomial')

EN_pred <- predict(new_EN_model, newx = new_x_test, type = 'class', s='lambda.min')

EN_confusion_matrix <- table(EN_pred, y_test)
print(EN_confusion_matrix)
```


- Precision of new EN model: 1
- Specificity of new EN model: 1
- Sensitivity of new EN model: 1


# 3.5 Build k-Nearest Neigbor model with filtered beta values
```{r eval=FALSE}
KNN_model <- knn(train = new_x_train, test = new_x_test, cl = y_train, k = 9)
KNN_confusion_matrix <- table(KNN_model, y_test)
print(KNN_confusion_matrix)
```

- Precision of KNN model: 1
- Specificity of KNN model: 1
- Sensitivity of KNN model: 1


# 3.6 Compare performance of machine learning models
```{r eval=FALSE}
comparison_ML_models <- cbind(EN_confusion_matrix, KNN_confusion_matrix)
```
<br></br>

# 4.1 Find relevant genes used to build EN model
```{r eval=FALSE}
new_EN_genes <- coef(new_EN_model, s='lambda.min')
new_EN_genes <- new_EN_genes[new_EN_genes[,1] != 0, ]
new_en_genes <- new_EN_genes[-1]
new_relevant_genes <- names(new_EN_genes) 

DNA_meth_gene_regions <- rowData(DNA_meth_data)
new_relevant_genes_names <- DNA_meth_gene_regions[new_relevant_genes, 'Gene_Symbol']
new_relevant_genes_names <- str_extract(new_relevant_genes_names, '[A-Za-z0-9]*')
```

# 4.2 Use hierarchical clustering to group samples by relevant genes
```{r eval=FALSE}
hmcol = colorRampPalette(rev(brewer.pal(9, "RdBu")))(256)
clust = function(x) hclust(x, method="complete")
dist = function(x) as.dist((1-cor(t(x)))/2)

#Interestingly, hierarchical clustering shows differential methylation of genes used by EN to classify samples

#If you develop problems getting the heatmap plot, check the new_relevant_genes. 
#Delete the Intercept value which may still be in the list
#new_relevant_genes <- new_relevant_genes[-1]
gene_heatmap = heatmap.2(
  t(beta_values[,new_relevant_genes]),
  scale="row",          # scale the values for each gene (row)
  density.info="none",  # turns off density plot inside color legend
  trace="none",         # turns off trace lines inside the heat map
  col=hmcol,            # define the color map
  labRow=new_relevant_genes_names, # use gene names instead of ensembl annotation
  labCol=FALSE,         # Not showing column labels
  ColSideColors=as.character(as.numeric(sample_types)), # Show colors for each response class
  dendrogram="both",    # Show dendrograms for both axis
  hclust = clust,       # Define hierarchical clustering method
  distfun = dist,       # Using correlation coefficient for distance function
  cexRow=0.86,            # Resize row labels
  margins=c(1,5)        # Define margin spaces
)
```

# 4.3 Survival analysis
```{r eval=FALSE}
clinical <- DNA_meth_data@colData
clin_df <- clinical[clinical$definition == 'Primary solid Tumor',
                    c('patient',
                      'vital_status',
                      'days_to_death',
                      'days_to_last_follow_up',
                      'gender',
                      'tumor_stage')]
clin_df$deceased = clin_df$vital_status == 'Dead'

clin_df$overall_survival = ifelse(clin_df$deceased,
                                  clin_df$days_to_death,
                                  clin_df$days_to_last_follow_up)


#check for differential methylation of the CDH4 gene
methyl_diseased <- beta_values[rownames(clin_df), relevant_genes[47]]
methyl_healthy <- beta_values[setdiff(rownames(beta_values), rownames(clin_df)), relevant_genes[47]]


boxplot(methyl_diseased, methyl_healthy, names = c('Diseased', 'Healthy'),
        main = 'Differential Methylation of 
        the CDH4 Gene between Breast Cancer and Normal Tissues')

```

# 4.4 Analysis on how degree of methylation affects survival
```{r eval=FALSE}
pval_list <- list() #list to store p-values of survival plots

for (i in 1:length(new_relevant_genes)){
  clin_df$beta_value <- beta_values[rownames(clin_df), new_relevant_genes[i]]
  median_value <- median(clin_df$beta_value)

  clin_df$meth <- ifelse(clin_df$beta_value >= median_value,
                        'HYPER',
                        'HYPO')


  fit <- survfit(Surv(overall_survival, deceased) ~ meth, data = clin_df)
  pval <- surv_pvalue(fit, data = clin_df)$pval
  pval_list <- append(pval_list, pval)
  png(paste(new_relevant_genes_names[i], '_plot', '.jpeg', sep = ''), width = 1000, height = 800)
  img <- ggsurvplot(fit, data = clin_df, pval = T, risk.table = T) #store image of survival plot
  
  print(img)
  
  dev.off()
  
}
```
